lib/tree.rb

```diff
@@ -12,44 +12,130 @@ def initialize(key, val)
 
 class Tree
   attr_reader :root
+  
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log n) if balanced and O(n) if not
+  # Space Complexity: O(log n) if balanced and O(n) if not
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, key, value)
+    # If we are already on nil, just add it here
+    if current.nil?
+      return TreeNode.new(key, value)
+    # Else if the key is less than, boop left 
+    elsif key < current.key
+      current.left = add_helper(current.left, key, value)
+    # Else if the key is less than, boop right 
+    elsif key > current.key
+      current.right = add_helper(current.right, key, value)
+    else
+    # Otherwise, we already have this value and we don't need to add it
+      raise Exception.new "This value has already been added" 
+    end 
+    
+    return current
+  end
+
+  # Time Complexity: O(log n) if balanced and O(n) if not
+  # Space Complexity: O(log n) if balanced and O(n) if not
   def find(key)
-    raise NotImplementedError
+    # If we are already on nil, just add it here
+    return nil if @root.nil?
+    #Return the root value if it is the value we are looking for
+    return @root.value if @root.key == key
+    #Otherwise, call in reinforcements
+    return find_helper(@root, key)
+  end
+  
+  def find_helper(current, key)
+    # If we are already on nil, just add it here
+    return nil if current.nil?
+    #if our current is now the same as the key, return it
+    return current.value if current.key == key
+    #Otherwise, if the key is less than the current key keep on booping left
+    if key < current.key
+      find_helper(current.left, key)
+    #If not, boop right
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) -- because we will always need to go throught ALL nodes
+  # Space Complexity: O(n) -- because we will always need to go throught ALL nodes
   def inorder
```

Comment:
👍 , but O(n) in space complexity because we are building an array of all the nodes.